ponycheck/generator.pony

```diff
@@ -1,68 +1,275 @@
+        match that.generate(rnd)
+        | let value_only: T => that.shrink(consume value_only)
+        | (let v: T, let shrinks: Iterator[T^]) => (consume v, consume shrinks)
+        end
+    end
+
+
 class box Generator[T] is GenObj[T]
+  """
+  A Generator is capable of generating random values of a certain type `T`
+  given a source of `Randomness`
+  and knows how to shrink or simplify values of that type.
+
+  When testing a property against one or more given Generators
+  those generators' `generate` methods are being called many times
+  to generate sample values that are then used to validate the property.
+
+  When a failing sample is found, the ponycheck engine is trying to find a
+  smaller or more simple sample by `shrinking` it.
+  If the generator did not provide any shrinked samples
+  as a result of `generate`, its `shrink` method is called
+  to obtain simpler results. Ponycheck obtains more shrunken samples until
+  the property is not failing anymore.
+  The last failing sample, which is considered the most simple one,
+  is then reported to the user.
+  """
   let _gen: GenObj[T]
 
   new create(gen: GenObj[T]) =>
     _gen = gen
 
-  fun generate(rnd: Randomness): T^ =>
+  fun generate(rnd: Randomness): GenerateResult[T] =>
+    """
+    Let this generator generate a value
+    given a source of `Randomness`.
+
+    Also allow for returning a value and pre-generated shrink results
+    as a `ValueAndShrink[T]` instance, a tuple of `(T^, Seq[T])`.
+    This helps propagating shrink results through all kinds of Generator
+    combinators like `filter`, `map` and `flatMap`.
+
+    If implementing a custom `Generator` based on another one,
+    a Generator Combinator, you should use shrunken values
+    returned by `generate` to also return shrunken values based on them
+
+    If generating an example value is costly, it might be more efficient
+    to simply return the generated value and only shrink in big steps or do no
+    shrinking at all.
+    If generating values is lightweight, shrunken values should also be returned.
+    """
     _gen.generate(rnd)
 
-  fun shrink(t: T): (T^, Seq[T]) =>
+  fun shrink(t: T): ValueAndShrink[T] =>
+    """
+    Simplify the given value.
+
+    As the returned value can also be `iso`, it needs to be consumed and returned
+
+    It is preffered to already return a `ValueAndShrink` from `generate`.
+    """
     _gen.shrink(consume t)
 
+  fun generate_value(rnd: Randomness): T^ =>
+    _gen.generate_value(rnd)
+
+  fun generate_and_shrink(rnd: Randomness): ValueAndShrink[T] =>
+    _gen.generate_and_shrink(rnd)
+
   fun filter(predicate: {(T): (T^, Bool)} box): Generator[T] =>
+    """
+    apply `predicate` to the values generated by this Generator
+    and only values for which `predicate` returns `true`.
+
+    Example:
+
+    ```pony
+    let even_i32s =
+      Generators.i32()
+        .filter(
+          {(t: I32}: (I32^, Bool) => (t % 2) == 0 })
```

Comment:
The lambda in this example doesn't seem to return a tuple

Reply:
my bad, corrected.